net/tsdial/tsdial.go

```diff
@@ -308,16 +312,25 @@ func (d *Dialer) userDialResolve(ctx context.Context, network, addr string) (net
 		return ipp, err
 	}
 
-	// Otherwise, hit the network.
-
 	// TODO(bradfitz): wire up net/dnscache too.
 
+	// Try tsdns resolver next to resolve SplitDNS
 	host, port, err := splitHostPort(addr)
 	if err != nil {
 		// addr is malformed.
 		return netip.AddrPort{}, err
 	}
 
+	if d.UserDialCustomResolver != nil {
+		ip, err := d.UserDialCustomResolver(host)
```

Comment:
Maybe instead of `UserDialCustomResolver` which will need to duplicate a bunch of the code below, this path could be an `else if` of the `exitDNSDoH` code path right below this, setting the `Dial` func to use the new connection to the internal resolver.

That should reduce the complexity in context handling.

I'm not sure what the best name would be, perhaps `UserDialCustomResolverDial`?

Reply:
Yes, that's better :) I reworked the code to use `UserDialCustomResolverDial` and set `r.dial = d.UserDialCustomResolverDial` when it is set, this should work properly and with the desired context